File: small_problems/medium_1/1000_lights.rb
=begin
PEDAC
  start 710
  finish 7:42
  time 32min
Problem
  rules
    accept an integer argument representing number of switches `n`
    toggle the switches on and off according to the following pattern
      switches start in the off position
      perform `n` iterations toggling switches
      on the 1st iteration 1, start at 1 and toggle every 1 switch
      on the 2nd iteration, start at 2 and toggle every 2nd switch
      on the 3rd iteration, start at 3 and toggle every 3rd switch
      etc
    return the number of every switch that is "on"
      switches are numbered 1 through n
  clarifying questions
    since 'n` represents a number of switches, assume input is always positive integer > 0?
      assume yes
Examples
  n = 5
                1 2 3 4 5 6 7 8 9 0
    Round 1 => [1 1 1 1 1 1 1 1 1 1]
    Round 2 => [1 0 1 0 1 0 1 0 1 0]
    Round 3 => [1 0 0 0 1 1 1 0 0 0]
    Round 4 => [1 0 0 1 1 1 1 1 0 0]
    Round 5 => [1 0 0 1 0 1 1 1 0 1]
    Round 6 => [1 0 0 1 0 0 1 1 0 1]
    Round 7 => [1 0 0 1 0 0 0 1 0 1]
    Round 8 => [1 0 0 1 0 0 0 0 0 1]
    Round 9 => [1 0 0 1 0 0 0 0 1 1]
    Round 10=> [1 0 0 1 0 0 0 0 1 0]
Data
  input: positive integer
  intermediate:
    booleans to represent on or off => true is on, false is off
    hash with integers as keys?
  output: array of (index + 1) for true values
Algorithm
  take argument `n` and create switches
    for each number 1-n, create a hash with number as key
    set switches to "off" position
      set keys equal to false
  for each number of switches (1 up to n)
    and for every multiple based on n (where `multiple` represents all multiples of curr switch num)
      get list of switches to toggle based on multiple (take current num up to the limit by num)
      for each of those, toggle that switch to off
  identify which switches are on and put them in an array
    select from hash where value == true
    select the keys for those values and put them in an array
  return the array
=end

def light_switch(switch_count)
  switches = {}
  1.upto(switch_count) { |num| switches[num] = false }

  1.upto(switch_count) do |switch|
    switch.step(to: switch_count, by: switch).each do |multiple|
      switches[multiple] = !switches[multiple]
    end
  end
  switches.select { |_, status| status }.keys
end

p light_switch(5) == [1, 4]
p light_switch(10) == [1, 4, 9]
p light_switch(1000) == [1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256,
                         289, 324, 361, 400, 441, 484, 529, 576, 625, 676, 729, 784, 841, 900, 961]

File: js_small_problems/interpretive_problem_solving/1000lights.js
/*
start: 604
end: 6:51
takeaways:
  - caught up in the proof,
  - not enough time in the examples and getting to a workable solution
  - distracted with solutions while trying to come up with questions
  - going through the motions on questions instead of thinking about what helps me  
*/

/*
Problem
  input: number identifying number of lights
  output: array of lights that are "on"
  rules:
    - accept an argument specifying the number of light switches
    - the argument also identifies the number of times to go through lights and toggle
    - for each pass, toggle every n lights, where n is the current pass through
      - example, for 3 lights:
        - first pass: toggle every 1 light starting with the first
        - second pass: toggle every 2nd light starting with the second
        - third pass: toggle eveyr 3rd light starting with the third
    - each light starts as 'off', each toggle goes from off to on
  questions
    inputs:
      - expected input: will it always be a number type integer greater than 0?
      - missing / extra inputs: should i expect missing inputs? how should i handle?
      - malformed inputs:
        - should i expect non Number types representing numbers (e.g. string '1')?
        - should i control for non numbers as input?
        - can there every be 0 or negative as an input? how should i handle?
      - extra inputs: can there be extra inputs?
      - others:
        - should i assume 1000 is the limit for number of lights?
    outputs:
      - should i always
Examples
  - good input:
    - even integer
    - odd integer
    - zero
  - malformed input:
Data Structure
  - input: integer
  - processing:
    - array of multiples
    - array of numbers 1 through n
  - output
Algorithm
  main program:
    - get array of numbers from 1 - n
      - [ 1, 2, 3, 4, 5]
      - [ 1, 2, 3, 4, 5, 6, 7, 8]
    - filter numbers for ones that have odd number of multiples from 1 - curr num
    - return filtered array
  filter callback function:
    - if multiplesOf % 2 !== 0
        1 -> true -> on
        2 -> false -> off
        3 -> true -> off
        4 -> false -> on
        5 -> false -> off
        6 -> false -> off
        7 -> false -> off
        8 -> false -> off
  multiplesOf
    - for each num from 1 to n
      - if n % eachNum === 0, push to result array
        1 -> [1]
        2 -> [1, 2]
        3 -> [1, 3]
        4 -> [1, 2, 4]
        5 -> [1, 5]
        6 -> [1, 2, 3, 6]
        7 -> [1, 7]
        8 -> [1, 2, 4, 8]
    - return result
Algorithm 2:
  build array of numbers 1 - sqrt of n truncated
  transform numbers into squared versions
  return squared verions
*/

function lightsOn(switches) {
  let limit = Math.trunc(Math.sqrt(switches));

  let initialNums = Array.from({ length: limit }, (_, idx) => (idx + 1) ** 2);
  console.log(initialNums);
}

lightsOn(5);        // [1, 4]
// Detailed result of each round for `5` lights
// Round 1: all lights are on
// Round 2: lights 2 and 4 are now off;     1, 3, and 5 are on
// Round 3: lights 2, 3, and 4 are now off; 1 and 5 are on
// Round 4: lights 2 and 3 are now off;     1, 4, and 5 are on
// Round 5: lights 2, 3, and 5 are now off; 1 and 4 are on

lightsOn(100);      // [1, 4, 9, 16, 25, 36, 49, 64, 81, 100]
lightsOn(6);
lightsOn(7);
lightsOn(8);
lightsOn(9);
lightsOn(10);
lightsOn(11);
lightsOn(12);
lightsOn(13);
lightsOn(14);
lightsOn(15);
lightsOn(16);